### PR TITLE
libonxx + 1.15.0 migration

### PR DESCRIPTION
**Pull request by hmaarrfk**
Note that the windows library is still static

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/0001-Link-to-abseil_dll.patch [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
-index 4dd56b6..3a002a5 100644
+index 0372c54b..4e1db018 100644
 --- a/CMakeLists.txt
 +++ b/CMakeLists.txt
-@@ -430,7 +430,7 @@ target_include_directories(onnx PUBLIC
+@@ -510,6 +510,7 @@ if(ONNX_USE_LITE_PROTO)
+ else()
+   if(TARGET protobuf::libprotobuf)
+     target_link_libraries(onnx_proto PUBLIC protobuf::libprotobuf PRIVATE ${protobuf_ABSL_USED_TARGETS})
++    target_link_libraries(onnx PUBLIC onnx_proto $ENV{PREFIX}/Library/lib/abseil_dll.lib)
```
Comment (hmaarrfk) [inline recipe/0001-Link-to-abseil_dll.patch:18]:
```suggestion
+    target_link_libraries(onnx_proto PUBLIC onnx_proto $ENV{PREFIX}/Library/lib/abseil_dll.lib)
```

File: recipe/0001-Link-to-abseil_dll.patch [hunk 2]
```diff
@@ -1,17 +1,25 @@
...
-index 4dd56b6..3a002a5 100644
+index 0372c54b..4e1db018 100644
 --- a/CMakeLists.txt
 +++ b/CMakeLists.txt
-@@ -430,7 +430,7 @@ target_include_directories(onnx PUBLIC
+@@ -510,6 +510,7 @@ if(ONNX_USE_LITE_PROTO)
+ else()
+   if(TARGET protobuf::libprotobuf)
+     target_link_libraries(onnx_proto PUBLIC protobuf::libprotobuf PRIVATE ${protobuf_ABSL_USED_TARGETS})
++    target_link_libraries(onnx_proto PUBLIC onnx_proto $ENV{PREFIX}/Library/lib/abseil_dll.lib)
```
Comment (hmaarrfk) [inline recipe/0001-Link-to-abseil_dll.patch:18]:
```suggestion
+    target_link_libraries(onnx_proto PUBLIC $ENV{PREFIX}/Library/lib/abseil_dll.lib)
```

**Comment by hmaarrfk** [comment]
I don't really understand what I am doing wrong on windows..... help appreciated.

**Comment by traversaro** [comment]
> I don't really understand what I am doing wrong on windows..... help appreciated.

`ONNX_API` is used as export macro for both `onnx_proto`  and `onnx`  libraries. So, `onnx_proto`  symbols get marked with `__declspec(dllexport)` also in `onnx`'s compilation units, and so the linker complains that they are not found even if they are present in the linked import librar `onnx_proto.lib` .

A quick fix (I just tested locally, it seems to work) is to change the export macro for `onnx-proto` :  https://github.com/traversaro/onnx/commit/e7719fd89c0bd12bc54983efeeeee5d926a07563 . It is not ideal as non-CMake downstream users needs to manually set `ONNX_PROTO_API` , but better than not working. A better fix is to define `ONNX_PROTO_API` in an header, but unfortunately protobuf does not support including an arbitrary header inclusion in the generated headers, so downstream projects typically do that via external scripts, see https://github.com/protocolbuffers/protobuf/issues/4198 for more details. If it is not a problem to expose all symbols of `onnx_proto` , an alternative simple solution is to set the `WINDOWS_EXPORT_ALL_SYMBOLS` CMake definition just for `onnx_proto`, and avoid to use any export pre-processor macro.

File: recipe/0001-Link-to-abseil_dll.patch [hunk 3]
```diff
@@ -1,17 +1,35 @@
...
+   if(TARGET protobuf::libprotobuf)
+     target_link_libraries(onnx_proto PUBLIC protobuf::libprotobuf PRIVATE ${protobuf_ABSL_USED_TARGETS})
++    target_link_libraries(onnx_proto PUBLIC $ENV{PREFIX}/Library/lib/abseil_dll.lib)
+   else()
+     target_link_libraries(onnx_proto PUBLIC ${PROTOBUF_LIBRARIES})
++    target_link_libraries(onnx_proto PUBLIC $ENV{PREFIX}/Library/lib/abseil_dll.lib)
+   endif()
+ endif()
+ add_onnx_global_defines(onnx_proto)
+@@ -528,7 +531,7 @@ target_include_directories(onnx PUBLIC
```
Comment (traversaro) [inline recipe/0001-Link-to-abseil_dll.patch:32]: If you find protobuf to `find_package(Protobuf CONFIG)`, this should be handled automatically by CMake.

Comment (traversaro) [reply recipe/0001-Link-to-abseil_dll.patch:32]: Anyhow, if this works fine it is ok to use that.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,63 +1,90 @@
...
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        - libprotobuf
+        - cmake
+        - make  # [unix]
+      host:
+        - libprotobuf
+        - libabseil  # [win]
```
Comment (traversaro) [inline recipe/meta.yaml:34]: I guess this should be present also on non-Windows platforms, so the correctly run_exports are exported for abseil (as it is a public dependency of libprotobuf).

Comment (hmaarrfk) [reply recipe/meta.yaml:34]: the linking analysis at the end doesn't make libaseil appear on unix

**Comment by hmaarrfk** [comment]
Thanks for the great explination.

Why is compilation not failing for https://github.com/conda-forge/onnx-feedstock/pull/103#discussion_r1383205503

File: recipe/bld_libonnx.bat [hunk 5]
```diff
@@ -0,0 +1,18 @@
+mkdir build
+cd build
+
+cmake %CMAKE_ARGS% ^
+    -DONNX_ML=1 ^
+    -DCMAKE_BUILD_TYPE=Release ^
+    -DCMAKE_CXX_STANDARD=17 ^
+    -DBUILD_SHARED_LIBS=ON ^
```
Comment (hmaarrfk) [inline recipe/bld_libonnx.bat:8]:
```suggestion
```

**Comment by hmaarrfk** [comment]
i see it is due to the fact that they aren't building shared libraries for windows.

File: recipe/meta.yaml (hunk 4 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:34]: i think this is due to the fact that linux uses shared libraries

**Comment by hmaarrfk** [comment]
I think `WINDOWS_EXPORT_ALL_SYMBOLS` is probably the best way to go.... but i'm not sure of any unintended effects.

**Comment by hmaarrfk** [comment]
maybe we can in CMake inject something into the header.

File: recipe/meta.yaml (hunk 4 above)
Comment (traversaro) [reply recipe/meta.yaml:34]: For reference, this was the discussion on libprotobuf now requiring linking to libabseil: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/4075#issuecomment-1562294904 .

**Comment by traversaro** [comment]
> maybe we can in CMake inject something into the header.

Just to understand, what is the goal now? Building as shared in Windows? In theory the patch in https://github.com/conda-forge/onnx-feedstock/pull/105#issuecomment-1794599952 should work fine, are you afraid the problem will create on non-cmake users of libonnx ?

**Comment by hmaarrfk** [comment]
> are you afraid the problem will create on non-cmake users of libonnx

yes.

I would ideally also like to upstream whatever patch i make since these build patches are pretty hard to maintain.

**Comment by hmaarrfk** [comment]
It seems onnx worked around this issue by creating an other header file.

Maybe I have to move files around to create this shared library

**Pull request closed without merging**